cluster-autoscaler/cloudprovider/kwok/kwok_provider.go

```diff
@@ -123,24 +125,28 @@ func (kwok *KwokCloudProvider) GetNodeGpuConfig(node *apiv1.Node) *cloudprovider
 
 // Refresh is called before every main loop and can be used to dynamically update cloud provider state.
 // In particular the list of node groups returned by NodeGroups can change as a result of CloudProvider.Refresh().
-// TODO(vadasambar): implement this
 func (kwok *KwokCloudProvider) Refresh() error {
 
-	// TODO(vadasambar): causes CA to not recognize kwok nodegroups
-	// needs better implementation
-	// nodeList, err := kwok.lister.List()
-	// if err != nil {
-	// 	return err
-	// }
+	allNodes, err := kwok.allNodesLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "failed to list all nodes from lister")
+		return err
+	}
+
+	targetSizeInCluster := make(map[string]int)
 
-	// ngs := []*NodeGroup{}
-	// for _, no := range nodeList {
-	// 	ng := parseAnnotationsToNodegroup(no)
-	// 	ng.kubeClient = kwok.kubeClient
-	// 	ngs = append(ngs, ng)
-	// }
+	for _, node := range allNodes {
+		ngName := getNGName(node, kwok.config)
+		if ngName == "" {
+			continue
+		}
 
-	// kwok.nodeGroups = ngs
+		targetSizeInCluster[ngName] += 1
+	}
+
+	for _, ng := range kwok.nodeGroups {
+		ng.targetSize = targetSizeInCluster[ng.Id()]
```

Comment:
Maybe something for the future: I wonder if we should delete nodes which have `ng` that is not a part of `kwok.nodeGroups`.